ui/src/ScholarReader.tsx

```diff
@@ -130,7 +130,14 @@ class ScholarReader extends React.PureComponent<ScholarReaderProps, State> {
   }
 
   setDrawerState(state: DrawerState) {
-    this.setState({ drawerState: state });
+    if (state === 'show-symbols') {
+      // There are a few edge cases where a close and show-symbol event 
+      // happen at the same time, thus we need to ensure the show-symbol is always 
+      // run last.
+      setTimeout(() => this.setState({ drawerState: state }), 0);
```

Comment:
I feel kind of uneasy about scheduling code to run with timers. Sometimes this is an indication that we should do a small code design elsewhere to make sure we get the intended behavior reliably with synchronous execution.

What are the edge cases? Can we refactor the code so they don't happen?

Reply:
Yes, I agree! Definitely doesn't sit right having to use setTimeout. The issue is that the click away listener listens to *any* clicks outside of the drawer container. This means that buttons that actually open the drawer end up sending close signals (due to the click away). I think this can be prevented by stoping bubbling within the button but I feel that's not very robust as it relies on any developer to know that if they wish to implement anything that opens the drawer.

Reply:
Ahhhh I see. Great point! What about adding an `onclick` listener to the document or page overlay that closes the drawer? Or, refactoring the `Drawer` so that instead of an explicit `open` property, it opens based on whether there is a symbol of citation selected? That way, once a symbol is deselected and the selected symbol is set to `null`, the drawer would automatically close?

Reply:
Ah I played a around a little bit with the onclick on the document but unfortunately end up running into the same issue about what should/shouldn't send close signals. The second idea sounds really cool tho - I'll take a crack at that!

Reply:
Hmmm so I've implemented this and realized that it unfortunately won't solve the issue :(. The problem is that the click away listener will listen to *any* clicks outside of the drawer. So in any system that has an external open mechanism to the drawer (regardless of the system) will always have this issue. This is because the sequence of event triggers will always be: opensignal (via some button) ==> close signal (picked up by a bubble to the document on the click away).... Not quite sure how to proceed

Reply:
I see, that makes sense. One workaround is to have the ClickAwayListener only attempt to dismiss the Drawer if a click is made on something other than an annotation, otherwise it keeps it open. Would that be too hacky?

Reply:
I see, I think this kind of circle's back to my other point that it's not very robust to start hard coding specific events that the drawer should know not to close for as then if someone wants to add another button that open's the drawer in the future, they need to also hard code that. It feels almost more hacky to me personally, but I am interested in what you think :o!

Reply:
Agreed. At this moment, I feel like whatever we choose will feel a _liiiitle_ hacky (which seems super weird because I feel like this interaction pattern is really common! It should be better supported by our programming tools.)

I'm thinking the alternative that sounds most appealing to me so far is to update the ClickAwayListener to not dismiss when other drawer-launching components are clicked, and to put clear comments both before the listener and at the top of the Drawer warning members of our team that if they want to add other drawer-launching components, they need to update the listener (e.g., by adding a new class to a list of classes that will be checked when determining whether to dismiss the drawer). I could be convinced that one of the other approaches we've discussed is cleaner however, though am coming to terms with the fact that there doesn't seem to be a fantastic solution to this XD

Reply:
Hahah okay, I'll move forward with this approach - but yeah it's a shame there isn't a cleaner solution for what would seem like a common interaction

Reply:
Okay just pushed a new solution that uses e.preventDefault(). Trying to see if a click came from a tooltip ended up being a little messy. Wondering what you think of this solution?

Reply:
Sorry to belabor this 😬 . What does the option for checking to see if a click came from a tooltip look like? The e.preventDefault() approach could work, though it does fragment click-away the logic into several different classes.

Reply:
It was really a combo of using typescript + trying to dig into the event. Since someone can click on the text within the button within the tooltip, I need to check if any parent in the DOM hierarchy of what was clicked has the class 'tooltip'. Which in and of itself is not a very robust solution imo bc now we are making the assumption that all tooltips will contain buttons that open the drawer.

Reply:
Maybe I can play around with that implementation to see if I can get the code to be a little cleaner? Although to be completely honest I think all of these solutions are bad and I'm starting to feel as though this may not be worth the time of re-implementing all these slightly better but also bad solutions hahahahah

Reply:
Fair enough. In that spirit, why we don't we go with your initial fix (e.preventDefault) for now and just clean it up. I'm thinking adding better breadcrumbs to describe the fix that will make the whole issue more easily understandable to anyone who has to maintain this in the future.

First, a hack-comment above the e.preventDefault() wherever it appears with a more thorough description of the design choice:

```typescript
/*
 * XXX(zkirby): ....
 */
```

And then another within the ClickAwayListener, emphasizing how the future code will have to be written to maintain the current behavior (i.e. that future annotations will also need a call to `e.preventDefault()`. How's that sound?

Reply:
@zkirby have you gotten a chance to incorporate this documentation into your PR?

Reply:
Yes! So sorry, committing now hahaha

Reply:
Well I found another really hairy bug while trying to finish the implementation on this :/. For some reason preventing event propagation in any part of the annotation (including it's own clickaway listener) prevents the annotation from closing which shouldn't happen. Not sure how to proceed from here, if we are committed to not using setTimeouts then it appears I'll have to implement your last suggestion with traversing the event chain and searching for a tooltip class :((((((

Reply:
Wow, what a bug! Can we do the traversing the even chain and searching for a tooltip class? I think it's pretty robust all things considered, and probably pretty easy to read and maintain as long as we put it in a nicely-named helper function.